File: pact-jvm-server/src/main/scala/au/com/dius/pact/server/Complete.scala
package au.com.dius.pact.server

import java.io.File
import au.com.dius.pact.core.model._
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._
import scala.util.Success

object Complete extends StrictLogging  {

  def getPort(j: Any): Option[String] = j match {
    case map: Map[AnyRef, AnyRef] => {
      if (map.contains("port")) Some(map("port").toString)
      else None
    }
    case _ => None
  }

  def toJson(error: VerificationResult) = {
    OptionalBody.body(("{\"error\": \"" + error + "\"}").getBytes)
  }

  def apply(request: Request, oldState: ServerState): Result = {
    def clientError = Result(new Response(400), oldState)
    def pactWritten(response: Response, port: String) = {
      val server = oldState(port)
      val newState = oldState.filter(p => p._2 != server)
      Result(response, newState)
    }

    val result = for {
      port <- getPort(JsonUtils.parseJsonString(request.getBody.valueAsString()))
      mockProvider <- oldState.get(port)
      sessionResults = mockProvider.getSession.remainingResults
      pact <- Option(mockProvider.getPact)
    } yield {
      mockProvider.stop()

      writeIfMatching(pact, sessionResults, mockProvider.getConfig.getPactVersion) match {
        case PactVerified => pactWritten(new Response(200, ResponseUtils.CrossSiteHeaders.asJava),
          mockProvider.getConfig.getPort.toString)
        case error => pactWritten(new Response(400,
          Map("Content-Type" -> List("application/json").asJava).asJava, toJson(error)),
          mockProvider.getConfig.getPort.toString)
      }
    }

    result getOrElse clientError
  }

  def writeIfMatching(pact: Pact, results: PactSessionResults, pactVersion: PactSpecVersion) = {
    if (results.allMatched) {
      val pactFile = destinationFileForPact(pact)
      DefaultPactWriter.INSTANCE.writePact(pactFile, pact, pactVersion)
    }
    VerificationResult(Success(results))
  }

  def defaultFilename[I <: Interaction](pact: Pact): String = s"${pact.getConsumer.getName}-${pact.getProvider.getName}.json"

  def destinationFileForPact[I <: Interaction](pact: Pact): File = destinationFile(defaultFilename(pact))

  def destinationFile(filename: String) = new File(s"${System.getProperty("pact.rootDir", "target/pacts")}/$filename")
}

File: pact-jvm-server/src/main/kotlin/au/com/dius/pact/server/Complete.kt
package au.com.dius.pact.server

import au.com.dius.pact.core.model.DefaultPactWriter
import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.Pact
import au.com.dius.pact.core.model.PactSpecVersion
import au.com.dius.pact.core.model.PactWriter
import au.com.dius.pact.core.model.Request
import au.com.dius.pact.core.model.Response
import au.com.dius.pact.core.support.Json
import au.com.dius.pact.core.support.json.JsonParser
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.File

private val logger = KotlinLogging.logger {}

object Complete {

  var pactWriter: PactWriter = DefaultPactWriter
  private val CrossSiteHeaders = mapOf("Access-Control-Allow-Origin" to listOf("*"))

  fun getPort(j: Any?): String? = when(j) {
    is Map<*, *> -> {
      if (j.contains("port")) j["port"].toString()
      else null
    }
    else -> null
  }

  fun toJson(error: VerificationResult) =
    OptionalBody.body(("{\"error\": \"$error\"}").toByteArray())

  fun parseJsonString(json: String?) =
    if (json == null || json.trim().isEmpty()) null
    else Json.fromJson(JsonParser.parseString(json))

  @JvmStatic
  fun apply(request: Request, oldState: ServerState): Result {
    fun pactWritten(response: Response, port: String) = run {
      val serverState = oldState.state
      val server = serverState[port]
      val newState = ServerState(oldState.state.filter { it.value != server })
      Result(response, newState)
    }

    val port = getPort(parseJsonString(request.body.valueAsString()))
    if (port != null) {
      val mockProvider = oldState.state[port]
      if (mockProvider != null) {
        val sessionResults = mockProvider.session.remainingResults()
        val pact = mockProvider.pact
        if (pact != null) {
          mockProvider.stop()

          return when (val result = writeIfMatching(pact, sessionResults, mockProvider.config.pactVersion)) {
            is VerificationResult.PactVerified -> pactWritten(Response(200, CrossSiteHeaders.toMutableMap()),
              mockProvider.config.port.toString())
            else -> pactWritten(Response(400, mapOf("Content-Type" to listOf("application/json")).toMutableMap(),
              toJson(result)), mockProvider.config.port.toString())
          }
        }
      }
    }

    return Result(Response(400), oldState)
  }

  fun writeIfMatching(pact: Pact, results: PactSessionResults, pactVersion: PactSpecVersion): VerificationResult {
    if (results.allMatched()) {
      val pactFile = destinationFileForPact(pact)
      pactWriter.writePact(pactFile, pact, pactVersion)
    }
    return VerificationResult.apply(au.com.dius.pact.core.support.Result.Ok(results))
  }

  fun defaultFilename(pact: Pact): String = "${pact.consumer.name}-${pact.provider.name}.json"

  fun destinationFileForPact(pact: Pact): File = destinationFile(defaultFilename(pact))

  fun destinationFile(filename: String) = File("${System.getProperty("pact.rootDir", "target/pacts")}/$filename")
}
